### ISSUE TITLE
key format and passphrase

### ISSUE THREAD

**Issue by rafipiccolo** [eCollect/node-ebics-client#1]
Hi, nice work

I would like to read an extract of information of my bank account from a script.
Is EBICS able to do that ?

I tried to do it from the information i got from https://www.certeurope.fr :
- I got a key :
I have a pem file (looks like an ssh public key)
and a der file (binary)
- I got some parameters : hostid, url, userid, partnerid.

Could you please give me some intel about your program so i can understand how it work.

- Can you explain what is the "passphrase" and what it looks like ?

- And what is the "key" format you want ? which is which

Thank you in advance for your reply

**Comment by vladhristov** [comment]
@rafipiccolo 
Hi Raphael,
This library implements EBICS standard, so if your bank supports it - you can use it to send orders - pay and get bank statements.

To use this library you would need:

**REGISTRATION:** Register with the bank. This will result in you having url, partnerId, userId, hostId.
**RSA KEYS:** Further more, to use this library and to be able to communicate with the bank you would need three RSA keys. These are not bank's, these are yours. If you do not have any, the library can generate those for you. The passphrase is a string of choice which is used to encrypt the RSA keys in the storage.

Then you have to start the initialisation process by sending three specific orders INI, HIA and HPB.
If successful, you can start sending payments and receive statements from the bank.

Does your bank have a test EBICS platform?
Regards,
Vlad

**Comment by rafipiccolo** [comment]
Hello, Thank you very much for your quick answer. It is very helpufull. I
will get back to you as soon as I succeed. In the meantime, I wish you a
good weekend

*Raphael Piccolo*
06.21.05.15.21
*rafi.piccolo@gmail.com <rafi.piccolo@gmail.com>*

**Comment by anandsahil** [comment]
Hi, 
Thank you for the library, i am a bit confused here. Please help me out.

I have all the URL , host , userId, customerId. From the above explanation i was able to generate keys.
Let me know if the sequence is wrong. This is what i am doing. I am integrating Zürcher Kantonalbank.

`const ebics = require("ebics-client")
const client = new ebics.Client({
		url: "simulate URl",
		partnerId: "customerId",
		userId: "userId",
		hostId: "hostId",
		passphrase: "test",
		keyStorage: ebics.fsKeysStorage("test1")
	}
);

client._generateKeys();
const iniOrder = ebics.Orders.INI;

client.initialization(INI).then(
	r => console.log("success", r)
).catch(
	t => console.log("error", t)
)
`
Thanks :-)

**Comment by vladhristov** [comment]
@anandsahil 
Hi,
First part, where you define clinet is OK.
You do not have to generate keys yourself. It is done automatically when initialisation process is started.
Here are the steps:
1. You create a clinet - you've done that.
2. You send initialisation process orders (INI and HIA) like that:
- get INI order - you've done it;
- with async/await: `const response = await client.send(iniOrder);`
- without async/await `client.send(iniOrder).then(r => console.log(r)).catch(e => console.log(e));`
- do the same for HIA order
3. You have to generate and print ini letter. You can do it like this:
```
const bankName = 'Zürcher Kantonalbank';
const template = fs.readFileSync('your_path_to_this_file/ini.hbs').toString();
const letter = new ebics.BankLetter({ client, bankName, template });

await letter.serialize('bankLetterFilename.html'); // or then, catch
```
- Open the html file. Print it. Send it to the bank. Now they are suppose to activate your account.

4. Once your account is activated by the bank, now you're ready to download their (bank) keys:
- Execute HPB order and get the keys
```
const hpbOrder = ebics.Orders.HPB;
const response  = await client.send(hpbOrder);
await client.setBankKeys(response.bankKeys);
```
5. Now you're ready to go.

Hope this will help.

Regards,
Vlad

**Comment by anandsahil** [comment]
@vladhristov thanks i will try it out how yo explained the steps.

Thank you :-)

**Comment by anandsahil** [comment]
@vladhristov this worked thank you, 
i also had a question, did you find any documentation for StandardOrderParams to send. 
For example is there a way to check that the file you uploaded is valid and did not have any errors via client ?

**Comment by nanov** [comment]
@anandsahil This really depends on your bank implementation, normally if there are errors you will receive an error message describing the error ( to what extents, again it really depends on your bank ).

You said you are working with ZKB, i would suggest that you first test your orders with their [test platform](https://testplattform.zkb.ch/help?locale=en) just to be on the safe side. It would be even better if you could report your results here so we can know about issues/compatibility.

**Comment by anandsahil** [comment]
@vladhristov i have some pull requests if you can look into it, would be really nice, i a develop might happen i would bring more  :-)

**Comment by nanov** [comment]
@anandsahil Thank you very much for your valuable contributions.

Your PRs have been merged and landed in 0.0.6, thank you!

**Comment by nanov** [comment]
As original topic of this issue is somehow gone into off-topic, and the rest of the concerns are addressed I am going to close this issue.

Please open new issues in case you experience any.
